text/0000-standard-lazy-types.md

```diff
@@ -0,0 +1,511 @@
+impl<T: UnwindSafe + RefUnwindSafe> RefUnwindSafe for OnceCell<T> {}
+```
+
+## Default type parameter on `Lazy`
+
+`Lazy` is defined with default type parameter.
+
+```rust
+pub struct Lazy<T, F = fn() -> T> { ... }
+```
+
+This is important to make using `Lazy` in static contexts convenient.
+Without this default, the user would have to type `T` type twice:
+
+```rust
+static GLOBAL_DATA: Lazy<Mutex<HashMap<i32, String>>, fn() -> Mutex<HashMap<i32, String>>
+    = Lazy::new(|| ... );
+```
+
+If we allow type inference in statics, this could be shortened to
+
+```rust
+static GLOBAL_DATA: Lazy<Mutex<HashMap<i32, String>>, _>
+    = Lazy::new(|| ... );
+```
+
+There are two drawbacks of using fn pointer type:
+
+* fn pointers are not ZSTs, so we waste one pointer per static lazy value.
+  Lazy locals will generally rely on type-inference and will use more specific closure type.
+* Specifying type for local lazy value might be tricky: `let x: Lazy<i32> = Lazy::new(|| closed_over_var)` fails with type error, the correct syntax is `let x: Lazy<i32, _> = Lazy::new(|| closed_over_var)`.
+
+## Only thread-safe flavor
+
+It is possible to add only `sync` version of the types, as they are the most useful.
+However, this would be against zero cost abstractions spirit.
+Additionally, non thread-safe version is required to replace `thread_local!` macro without imposing synchronization.
+
+## Synchronization Guarantees
+
+In theory, it is possible to specify two different synchronization guarantees for `get` operation, release/acquire or release/consume.
+They differ in how they treat side effects.
+If thread **A** executes `get_or_init(f)`, and thread **B** executes `get` and observes the value, release/acquire guarantees that **B** also observes side-effects of `f`.
+
+Here's a program which allows to observe the difference:
+
+```rust
+static FLAG: AtomicBool = AtomicBool::new(false);
+static CELL: OnceCell<()> = OnceCell::new();
+
+// thread1
+CELL.get_or_init(|| FLAG.store(true, Relaxed));
+
+// thread2
+if CELL.get().is_some() {
+  assert!(FLAG.load(Relaxed))
+}
+```
+
+Under release/acquire, the assert never fires.
+Under release/consume, it might fire.
+
+Release/consume can potentially be implemented more efficiently on weak memory model architectures.
+However, the situation with `consume` ordering is cloudy right now:
+
+* [nobody knows what it actually means](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2016/p0371r0.html),
+* [but people rely on it in practice for performance](https://docs.rs/crossbeam-utils/0.7.0/crossbeam_utils/atomic/trait.AtomicConsume.html#tymethod.load_consume).
+
+We can do one of the following:
+
+1. Specify and implement `acquire` ordering,
+2. Specify `consume` but implement `acquire` (or hack `consume` in an implementation-defined manner) with the hope to make implementation more efficient later.
+3. Specify and implement `acquire`, but provide additional API which can take `Ordering` as an argument.
```

Comment:
> The exact definition of "depend on" is a bit vague, but it works as you would expect in practice since a lot of software, especially the Linux kernel, rely on this behavior.

- crossbeam_utils::AtomicConsume::load_consume

> It is widely accepted that the current definition of memory_order_consume in the standard is not useful. All current compilers essentially map it to memory_order_acquire. 
...
>  Currently there is no way to express this code in a way that is both performant and standard-conforming.
...
> It is likely that any such revision will not be fully backwards-compatible with the old one, but will better support common use cases. 

- the deprecation proposal for memory_order_consume

The links are helpful, but examining them raises more confusion. And some architectures Rust supports default to acquire-based semantics, so someone compiling on them would not expect the bugs that arise in another architecture because they always acquire. Others default to consume-like semantics, and run slower if forced to acquire. However, because of the former, it seems like embedding an implicit consume contract is an invitation for Heisenbugs because a user cannot actually know and try to test for if the access is consume-ordered or not, no?

Reply:
> However, because of the former, it seems like embedding an implicit consume contract is an invitation for Heisenbugs because a user cannot actually know and try to test for if the access is consume-ordered or not, no?

This is true more or less about all atomics in Rust: most people test only on x86 (I certainly do), and it implements a relatively strong memory model.

Reply:
That's true, but a lot of the `atomic` APIs accept `atomic::Ordering`, and Rust doesn't expose a variant of `atomic::Ordering::Consume`, so it seems premature to specify it as a default.

And if someone writes code that works on one architecture but is different on another, well, arguably that's the breaks. It's rust-lang not miracle-lang.  But specifying Consume, implementing Acquire, and then on _one_ architecture **relaxing** to Consume when `atomic::Ordering::Consume` becomes available seems like deliberately breaking the API for one architecture in a way that may not be easily understandable, and testing is hard even if you have such a machine. So I suppose I will narrow it down further: specifying Consume as a default **could** be OK, but should that method remain unstable until an Ordering::Consume is also stabilized? I am not as sure about 2b, since it seems like a very different case in actuality.

I'm in favor of standard lazy types, it just seemed like something to think about.

Reply:
On the upside, since I made this comment, a bunch of aarch64 targets arrived in tier 1 and tier 2 with host tools, and at least one is even a popular developer platform, so debugging this admittedly seems more plausible.